plugins/inputs/ping/ping.go

```diff
@@ -194,137 +177,153 @@ func hostPinger(binary string, timeout float64, args ...string) (string, error)
+			rsps = append(rsps, res)
+		}
+		r.Done()
+	}()
+
+	packetsSent := 0
+	wg := &sync.WaitGroup{}
+	c := ping.Client{}
+
+	for p.Count <= 0 || packetsSent < p.Count {
+		select {
+		case <-ctx.Done():
+			goto finish
+		case <-tick.C:
+			ctx, cancel := context.WithTimeout(context.Background(), time.Duration(timeout*float64(time.Second)))
+			defer cancel()
+
+			packetsSent++
+			wg.Add(1)
+			go func(seq int) {
+				defer wg.Done()
+				resp, err := c.Do(ctx, ping.Request{
+					Dst: net.ParseIP(host.String()),
+					Src: net.ParseIP(p.listenAddr),
+					Seq: seq,
+				})
+				if err != nil {
+					// likely a timeout error, ignore
+					return
+				}
+
+				resps <- resp
+			}(packetsSent)
 		}
 	}
-	return trans, recv, ttl, min, avg, max, stddev, err
+
+finish:
+	wg.Wait()
+	close(resps)
+
+	r.Wait()
+	tags, fields := onFin(packetsSent, rsps, destination)
+	acc.AddFields("ping", fields, tags)
 }
 
-func getPacketStats(line string, trans, recv int) (int, int, error) {
-	stats := strings.Split(line, ", ")
-	// Transmitted packets
-	trans, err := strconv.Atoi(strings.Split(stats[0], " ")[0])
-	if err != nil {
-		return trans, recv, err
+func onFin(packetsSent int, resps []*ping.Response, destination string) (map[string]string, map[string]interface{}) {
+	packetsRcvd := len(resps)
+	loss := float64(packetsSent-packetsRcvd) / float64(packetsSent) * 100
+
+	tags := map[string]string{"url": destination}
+	fields := map[string]interface{}{
+		"result_code":         0,
+		"packets_transmitted": packetsSent,
+		"packets_received":    packetsRcvd,
+		"percent_packet_loss": loss,
 	}
-	// Received packets
-	recv, err = strconv.Atoi(strings.Split(stats[1], " ")[0])
-	return trans, recv, err
-}
 
-func getTTL(line string) (int, error) {
-	ttlLine := regexp.MustCompile(`ttl=(\d+)`)
-	ttlMatch := ttlLine.FindStringSubmatch(line)
-	return strconv.Atoi(ttlMatch[1])
-}
+	if packetsRcvd == 0 || packetsSent == 0 {
+		return tags, fields
+	}
 
-func checkRoundTripTimeStats(line string, min, avg, max,
-	stddev float64) (float64, float64, float64, float64, error) {
-	stats := strings.Split(line, " ")[3]
-	data := strings.Split(stats, "/")
+	ttl := resps[0].TTL
 
-	min, err := strconv.ParseFloat(data[0], 64)
-	if err != nil {
-		return min, avg, max, stddev, err
+	var min, max, avg, total time.Duration
+	min = resps[0].RTT
+	max = resps[0].RTT
+
+	for _, res := range resps {
+		if res.RTT < min {
+			min = res.RTT
+		}
+		if res.RTT > max {
+			max = res.RTT
+		}
+		total += res.RTT
 	}
-	avg, err = strconv.ParseFloat(data[1], 64)
-	if err != nil {
-		return min, avg, max, stddev, err
+
+	avg = total / time.Duration(packetsRcvd)
+	var sumsquares time.Duration
+	for _, res := range resps {
+		sumsquares += (res.RTT - avg) * (res.RTT - avg)
 	}
-	max, err = strconv.ParseFloat(data[2], 64)
-	if err != nil {
-		return min, avg, max, stddev, err
+	stdDev := time.Duration(math.Sqrt(float64(sumsquares / time.Duration(packetsRcvd))))
+	if ttl > 0 {
```

Comment:
Yes, it was to mirror the exec ping, if they were 0, it would likely have returned before this. Does it make sense to keep checking and prevent `ttl` from showing up in windows (it will be 0 until it's supported).

Reply:
I think in the original code we returned these stats from a function as non-pointer values, and lose the ability to tell the difference between "don't know" and 0.  This seems like a minor bug in the old code.

In general, when we don't know what one of the values is we don't want to set it and if we do know, even if its zero, we should.  Here I think we have 2 cases: TTL doesn't work on Windows and TTL is legitimately 0 (not sure if this is possible, maybe it can only go to 1, but doesn't actually matter).

If the TTL isn't supported in Windows, we should simply check if we are on Windows and not report TTL.  This is a more direct way of dealing with shortcoming that are platform specific; add a comment that TTL is not supported on Windows with the go icmp package.

If the TTL is zero, I think we just report it, same goes for min/max/mean/stddev, zero seems valid though unlikely for all of these.